pexec/managed_process_windows.go

```diff
@@ -35,75 +37,117 @@ func (p *managedProcess) sysProcAttr() (*syscall.SysProcAttr, error) {
-			return false, nil
+			return false, &ProcessNotExistsError{err}
 		default:
 			return false, errors.Wrapf(err, "error killing process %d", p.cmd.Process.Pid)
 		}
 	}
 
-	if !shouldJustForce {
-		// In case the process didn't stop, or left behind any orphan children in its process group,
-		// we now ask everything in the process tree to stop after a brief wait.
-		timer := time.NewTimer(p.stopWaitInterval)
-		defer timer.Stop()
-		select {
-		case <-timer.C:
-			p.logger.Infof("killing entire process tree %d", p.cmd.Process.Pid)
-			if out, err := exec.Command("taskkill", "/t", "/pid", pidStr).CombinedOutput(); err != nil {
-				switch {
-				case strings.Contains(string(out), mustForce):
-					p.logger.Debug("must force terminate process tree")
-					shouldJustForce = true
-				case strings.Contains(string(out), "not found"):
-					return false, nil
-				default:
-					return false, errors.Wrapf(err, "error killing process tree %d", p.cmd.Process.Pid)
-				}
+	// In case the process didn't stop, or left behind any orphan children in its process group,
+	// we now ask everything in the process tree to stop after a brief wait.
+	timer := time.NewTimer(p.stopWaitInterval)
+	defer timer.Stop()
+	select {
+	case <-timer.C:
+		p.logger.Infof("killing entire process tree %d", p.cmd.Process.Pid)
+		out, err := exec.Command("taskkill", "/t", "/pid", pidStr).CombinedOutput()
+		if err != nil {
+			switch {
+			case strings.Contains(string(out), mustForce):
+				p.logger.Debug("must force terminate process tree")
+				shouldJustForce = true
+			case strings.Contains(string(out), "not found"):
+				return false, &ProcessNotExistsError{err}
+			default:
+				return false, errors.Wrapf(err, "error killing process tree %d", p.cmd.Process.Pid)
 			}
-		case <-p.managingCh:
-			timer.Stop()
 		}
+	case <-p.managingCh:
+		timer.Stop()
 	}
 
 	// Lastly, kill everything in the process tree that remains after a longer wait or now. This is
 	// going to likely result in an "exit status 1" that we will have to interpret.
 	// FUTURE(erd): find a way to do this better. Research has not come up with much and is
 	// program dependent.
-	var forceKilled bool
-	if !shouldJustForce {
-		timer2 := time.NewTimer(p.stopWaitInterval * 2)
-		defer timer2.Stop()
+
+	// We can force kill the process group right away, if the flag is already set
+	forceKillCommand := exec.Command("taskkill", "/t", "/f", "/pid", pidStr)
+	if shouldJustForce {
 		select {
-		case <-timer2.C:
-			p.logger.Infof("force killing entire process tree %d", p.cmd.Process.Pid)
-			if err := exec.Command("taskkill", "/t", "/f", "/pid", pidStr).Run(); err != nil {
+		case <-p.managingCh:
+			return false, nil
+		default:
+			if err := forceKillCommand.Run(); err != nil {
 				return false, errors.Wrapf(err, "error force killing process tree %d", p.cmd.Process.Pid)
 			}
-			forceKilled = true
-		case <-p.managingCh:
-			timer2.Stop()
+			return true, nil
 		}
```

Comment:
We were already waiting on the management channel in the previous section so I wouldn't even bother checking it here. You can just skip straight to the code currently in the `default` case.